File: app/src/main/java/org/blitzortung/android/alert/AlertResult.java
package org.blitzortung.android.alert;

import org.blitzortung.android.alert.data.AlertSector;

public class AlertResult {
	
	private final AlertSector sector;
    
    private final String distanceUnitName;

    public AlertResult(AlertSector sector, String distanceUnitName) {
		this.sector = sector;
        this.distanceUnitName = distanceUnitName;
    }

	public float getClosestStrikeDistance() {
		return sector.getClosestStrikeDistance();
	}

    public String getDistanceUnitName() {
        return distanceUnitName;
    }

    public String getBearingName() {
        return sector.getLabel();
    }
    
    @Override
    public String toString() {
        return String.format("%s %.1f %s", getBearingName(), getClosestStrikeDistance(), getDistanceUnitName());
    }
}

File: app/src/main/java/org/blitzortung/android/alert/AlertResult.kt
package org.blitzortung.android.alert

import org.blitzortung.android.alert.data.AlertSector

data class AlertResult(
        val sectors: List<AlertSector>,
        val parameters: AlertParameters,
        val referenceTime: Long
) {
    val sectorsByDistance: Map<Float, AlertSector> by lazy {
        sectors.filter { it.closestStrikeDistance < Float.POSITIVE_INFINITY }
                .sortedBy { it.closestStrikeDistance }
                .toMapBy { it.closestStrikeDistance }
    }

    val sectorWithClosestStrike: AlertSector? by lazy {
        sectorsByDistance.values.firstOrNull()
    }

    val closestStrikeDistance: Float
        get() = sectorWithClosestStrike?.closestStrikeDistance ?: Float.POSITIVE_INFINITY

    val bearingName: String
        get() = sectorWithClosestStrike?.label ?: "n/a"

    override fun toString(): String {
        return "%s %.1f %s".format(bearingName, closestStrikeDistance, parameters.measurementSystem.unitName)
    }
}
